tests/two_d_array_ut/two_d_array_ut.c

```diff
@@ -0,0 +1,642 @@
+/* Tests_SRS_TWO_D_ARRAY_07_006: [ TWO_D_ARRAY_CREATE(T) shall succeed and return a non-NULL value. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_CREATE_create_uint32_t_with_bigger_size_succeeds)
+{
+    //arrange
+    STRICT_EXPECTED_CALL(malloc_flex(IGNORED_ARG, 10000000, sizeof(uint32_t*)));
+
+    //act
+    TWO_D_ARRAY(uint32_t) tdarr = TWO_D_ARRAY_CREATE(uint32_t)(10000000, 5000);
+
+    //assert
+    ASSERT_IS_NOT_NULL(tdarr);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    //clean
+    TWO_D_ARRAY_ASSIGN(uint32_t)(&tdarr, NULL);
+}
+
+/* Tests_SRS_TWO_D_ARRAY_07_003: [ TWO_D_ARRAY_CREATE(T) shall call THANDLE_MALLOC_FLEX with TWO_D_ARRAY_FREE(T) as dispose function, nmemb set to row_size and size set to sizeof(T*). ]*/
+/* Tests_SRS_TWO_D_ARRAY_07_004: [ TWO_D_ARRAY_CREATE(T) shall set all rows point to NULL. ]*/
+/* Tests_SRS_TWO_D_ARRAY_07_006: [ TWO_D_ARRAY_CREATE(T) shall succeed and return a non-NULL value. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_CREATE_create_THANDLE_succeeds)
+{
+    //arrange
+    STRICT_EXPECTED_CALL(malloc_flex(IGNORED_ARG, 5, sizeof(THANDLE(A_TEST)*)));
+
+    //act
+    TWO_D_ARRAY(THANDLE(A_TEST)) tdarr = TWO_D_ARRAY_CREATE(THANDLE(A_TEST))(5, 5);
+
+    //assert
+    ASSERT_IS_NOT_NULL(tdarr);
+    for (unsigned int i = 0; i < 5; i++)
+    {
+        ASSERT_IS_NULL(tdarr->row_arrays[i]);
+    }
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    //clean
+    TWO_D_ARRAY_ASSIGN(THANDLE(A_TEST))(&tdarr, NULL);
+}
+
+/* Tests_SRS_TWO_D_ARRAY_07_003: [ TWO_D_ARRAY_CREATE(T) shall call THANDLE_MALLOC_FLEX with TWO_D_ARRAY_FREE(T) as dispose function, nmemb set to row_size and size set to sizeof(T*). ]*/
+/* Tests_SRS_TWO_D_ARRAY_07_004: [ TWO_D_ARRAY_CREATE(T) shall set all rows point to NULL. ]*/
+/* Tests_SRS_TWO_D_ARRAY_07_006: [ TWO_D_ARRAY_CREATE(T) shall succeed and return a non-NULL value. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_CREATE_create_THANDLE_with_bigger_size_succeeds)
+{
+    //arrange
+    STRICT_EXPECTED_CALL(malloc_flex(IGNORED_ARG, 10000000, sizeof(THANDLE(A_TEST)*)));
+
+    //act
+    TWO_D_ARRAY(THANDLE(A_TEST)) tdarr = TWO_D_ARRAY_CREATE(THANDLE(A_TEST))(10000000, 5000);
+
+    //assert
+    ASSERT_IS_NOT_NULL(tdarr);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    //clean
+    TWO_D_ARRAY_ASSIGN(THANDLE(A_TEST))(&tdarr, NULL);
+}
+
+/* Tests_SRS_TWO_D_ARRAY_07_005: [ If there are any errors then TWO_D_ARRAY_CREATE(T) shall fail and return NULL. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_CREATE_fails_when_underlying_calls_fails)
+{
+    //arrange
+    STRICT_EXPECTED_CALL(malloc_flex(IGNORED_ARG, 5, sizeof(uint32_t*)))
+        .SetReturn(NULL);
+
+    //act
+    TWO_D_ARRAY(uint32_t) tdarr = TWO_D_ARRAY_CREATE(uint32_t)(5, 5);
+
+    //assert
+    ASSERT_IS_NULL(tdarr);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    //clean
+}
```

Comment:
Consider using umock_negative_tests
We usually use that as a general pattern in all the when_underlying_calls_fails_XXX_api_fails

Reply:
changed to use umock_c_negative_test pattern